File: sdk-dotnet/AriaBsv.Core/InferenceAuditor.cs
// aria-bsv/dotnet — Client-side inference auditor (BRC-121).
//
// Mirrors the Python SDK's aria.core.auditor module:
//   - Record()   hashes input + output and appends to the current batch.
//   - Flush()    finalises the epoch: computes the Merkle root over all records
//                and fires OnFlush. Does NOT broadcast to BSV (BSV integration
//                is the responsibility of the host application).
//   - Reset()    clears all state and starts a fresh epoch.
//
// Raw inputs and outputs are NEVER stored — only their SHA-256 hashes.

namespace AriaBsv.Core;

/// <summary>
/// Snapshot of auditor counters returned by <see cref="InferenceAuditor.Stats"/>.
/// </summary>
/// <param name="TotalRecords">Total records created since construction.</param>
/// <param name="TotalFlushes">Number of successful flushes performed.</param>
/// <param name="CurrentBatchSize">Records pending in the current unflushed batch.</param>
public record AuditStats(int TotalRecords, int TotalFlushes, int CurrentBatchSize);

/// <summary>
/// Records AI inferences and batches them into verifiable epochs anchored by
/// a Merkle root (BRC-121 protocol).
/// </summary>
/// <example>
/// <code>
/// var auditor = new InferenceAuditor("my-ai-system", batchSize: 100);
/// string id = auditor.Record("gpt-4o", inputObj, outputObj, 0.97, 42);
/// var (root, records) = auditor.Flush();
/// </code>
/// </example>
public sealed class InferenceAuditor
{
    private readonly string _systemId;
    private readonly int _batchSize;

    private string _epochId;
    private int _epochSeq;
    private MerkleTree _tree;
    private readonly List<AuditRecord> _records = new();

    private int _totalRecords;
    private int _totalFlushes;

    // -------------------------------------------------------------------------
    // Events
    // -------------------------------------------------------------------------

    /// <summary>
    /// Fired at the end of every successful <see cref="Flush"/> with
    /// the computed Merkle root and the flushed records.
    /// </summary>
    public event Action<string, IReadOnlyList<AuditRecord>>? OnFlush;

    // -------------------------------------------------------------------------
    // Construction
    // -------------------------------------------------------------------------

    /// <summary>
    /// Create an auditor for <paramref name="systemId"/>.
    /// </summary>
    /// <param name="systemId">Identifier of the audited AI system.</param>
    /// <param name="batchSize">
    /// Auto-flush threshold: when the pending batch reaches this size,
    /// <see cref="Flush"/> is called automatically. 0 means no auto-flush.
    /// </param>
    public InferenceAuditor(string systemId, int batchSize = 0)
    {
        _systemId  = systemId;
        _batchSize = batchSize;
        _epochId   = GenerateEpochId();
        _tree      = new MerkleTree();
    }

    // -------------------------------------------------------------------------
    // Properties
    // -------------------------------------------------------------------------

    /// <summary>Current epoch identifier.</summary>
    public string EpochId => _epochId;

    /// <summary>Auditor counters.</summary>
    public AuditStats Stats => new(_totalRecords, _totalFlushes, _records.Count);

    // -------------------------------------------------------------------------
    // Recording
    // -------------------------------------------------------------------------

    /// <summary>
    /// Record a single AI inference with no caller metadata.
    /// Raw inputs and outputs are hashed; the original values are not stored.
    /// </summary>
    /// <returns>The unique record ID for this entry.</returns>
    public string Record(
        string modelId,
        object? input,
        object? output,
        double? confidence,
        long latencyMs)
        => RecordWithMeta(modelId, input, output, confidence, latencyMs,
                          new Dictionary<string, object?>());

    /// <summary>
    /// Record a single AI inference with additional caller-supplied metadata.
    /// </summary>
    /// <returns>The unique record ID for this entry.</returns>
    public string RecordWithMeta(
        string modelId,
        object? input,
        object? output,
        double? confidence,
        long latencyMs,
        Dictionary<string, object?> metadata)
    {
        var inputHash  = Hasher.HashObject(input);
        var outputHash = Hasher.HashObject(output);

        var seq      = _epochSeq++;
        var recordId = $"rec_{_epochId}_{seq:D6}";
        var ts       = DateTime.UtcNow.ToString("o");

        var record = new AuditRecord(
            RecordId:   recordId,
            SystemId:   _systemId,
            ModelId:    modelId,
            InputHash:  inputHash,
            OutputHash: outputHash,
            Confidence: confidence,
            LatencyMs:  latencyMs,
            Timestamp:  ts,
            Sequence:   seq,
            EpochId:    _epochId,
            Metadata:   metadata);

        _records.Add(record);
        _totalRecords++;

        // Hash the record's canonical representation and add to the Merkle tree.
        // Uses CanonicalJson so that the leaf hash is cross-SDK reproducible.
        var recordJson = Hasher.CanonicalJson(RecordToDict(record));
        _tree.AddLeaf(recordJson);

        // Auto-flush when batch size is reached
        if (_batchSize > 0 && _records.Count >= _batchSize)
            Flush();

        return recordId;
    }

    // -------------------------------------------------------------------------
    // Flush / Reset
    // -------------------------------------------------------------------------

    /// <summary>
    /// Finalise the current batch: compute the Merkle root, fire
    /// <see cref="OnFlush"/>, reset state, and return the root + records.
    /// Returns an empty root string when the batch is empty.
    /// </summary>
    public (string MerkleRoot, IReadOnlyList<AuditRecord> Records) Flush()
    {
        if (_records.Count == 0)
            return (string.Empty, Array.Empty<AuditRecord>());

        var merkleRoot = _tree.Root();
        var flushedRecords = _records.AsReadOnly();

        _totalFlushes++;
        OnFlush?.Invoke(merkleRoot, flushedRecords);

        // Keep a snapshot before resetting
        var snapshot = _records.ToList().AsReadOnly();
        Reset();
        return (merkleRoot, snapshot);
    }

    /// <summary>
    /// Clear all pending records, reset the Merkle tree, and start a new epoch.
    /// Does NOT increment the flush counter.
    /// </summary>
    public void Reset()
    {
        _records.Clear();
        _tree     = new MerkleTree();
        _epochId  = GenerateEpochId();
        _epochSeq = 0;
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    /// <summary>
    /// Convert an <see cref="AuditRecord"/> to a snake_case dictionary that
    /// matches the serialisation used by the Python and TypeScript SDKs.
    /// </summary>
    private static Dictionary<string, object?> RecordToDict(AuditRecord r) => new()
    {
        ["record_id"]   = r.RecordId,
        ["system_id"]   = r.SystemId,
        ["model_id"]    = r.ModelId,
        ["input_hash"]  = r.InputHash,
        ["output_hash"] = r.OutputHash,
        ["confidence"]  = r.Confidence,    // double? → null or number
        ["latency_ms"]  = r.LatencyMs,     // long → integer
        ["timestamp"]   = r.Timestamp,
        ["sequence"]    = (object)r.Sequence,  // int → integer
        ["epoch_id"]    = r.EpochId,
        ["metadata"]    = r.Metadata,
    };

    private static string GenerateEpochId()
    {
        var ts   = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var rand = Random.Shared.Next(0, 0xFFFFFF).ToString("x6");
        return $"{ts}_{rand}";
    }
}

File: sdk-java/src/main/java/io/aria/bsv/auditor/InferenceAuditor.java
package io.aria.bsv.auditor;

import com.google.gson.JsonObject;
import io.aria.bsv.hasher.Hasher;
import io.aria.bsv.merkle.MerkleException;
import io.aria.bsv.merkle.MerkleTree;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * Records AI inference events and produces BRC-121 audit epochs.
 *
 * <p>Usage pattern:
 * <ol>
 *   <li>Construct with a system identifier.</li>
 *   <li>Call {@link #record} for every inference event.</li>
 *   <li>Call {@link #closeEpoch} to seal the batch, receive the Merkle root,
 *       and reset the auditor for the next epoch.</li>
 * </ol>
 *
 * <p>Raw inputs and outputs are never retained — only their SHA-256 hashes
 * (via {@link Hasher#hashObject}) are stored in each {@link AuditRecord}.
 */
public final class InferenceAuditor {

    private final String systemId;

    /** Current epoch identifier; reset on each {@link #closeEpoch} call. */
    private String epochId;

    /** Accumulated records for the current epoch. */
    private final List<AuditRecord> records = new ArrayList<>();

    /** 0-based sequence counter; reset on each {@link #closeEpoch} call. */
    private int sequence = 0;

    /**
     * Create a new auditor for the given system.
     *
     * @param systemId stable identifier for the AI system being audited
     *                 (e.g. {@code "my-recommender-v2"})
     */
    public InferenceAuditor(String systemId) {
        this.systemId = systemId;
        this.epochId  = UUID.randomUUID().toString();
    }

    // -------------------------------------------------------------------------
    // Recording
    // -------------------------------------------------------------------------

    /**
     * Record a single inference event.
     *
     * <p>The method hashes {@code input} and {@code output} via
     * {@link Hasher#hashObject} (canonical JSON, then SHA-256) and stores the
     * resulting {@link AuditRecord} internally for the current epoch.
     *
     * @param modelId human-readable model identifier (e.g. {@code "gpt-4o"})
     * @param input   raw inference input as a Gson {@link JsonObject}
     * @param output  raw inference output as a Gson {@link JsonObject}
     * @return the created {@link AuditRecord} (caller may inspect or log it)
     */
    public AuditRecord record(String modelId, JsonObject input, JsonObject output) {
        String inputHash  = Hasher.hashObject(input);
        String outputHash = Hasher.hashObject(output);

        String timestamp = Instant.now().toString(); // ISO-8601 UTC
        String recordId  = String.format("rec_%s_%06d", epochId, sequence);

        AuditRecord rec = new AuditRecord(
                recordId,
                systemId,
                modelId,
                inputHash,
                outputHash,
                null,                   // confidence — not measured here
                0L,                     // latencyMs  — not measured here
                timestamp,
                sequence,
                epochId,
                Collections.emptyMap()
        );

        records.add(rec);
        sequence++;
        return rec;
    }

    // -------------------------------------------------------------------------
    // Epoch management
    // -------------------------------------------------------------------------

    /**
     * Close the current epoch.
     *
     * <p>Computes a {@link MerkleTree} over the SHA-256 hashes of all records
     * accumulated since the last close (or since construction), seals the epoch
     * summary, then resets the auditor so a new epoch can begin.
     *
     * <p>Each Merkle leaf is the UTF-8 encoding of the hex digest produced by
     * {@link Hasher#hashObject} on the record's canonical map.
     *
     * @return {@link EpochSummary} containing the epoch identifier, system
     *         identifier, record count, and Merkle root
     * @throws MerkleException if no records have been added to the current epoch
     */
    public EpochSummary closeEpoch() throws MerkleException {
        MerkleTree tree = new MerkleTree();
        for (AuditRecord rec : records) {
            String recHash = Hasher.hashObject(rec.toCanonicalMap());
            tree.addLeaf(recHash.getBytes(StandardCharsets.UTF_8));
        }

        String merkleRoot = tree.root(); // throws MerkleException if no records

        EpochSummary summary = new EpochSummary(
                epochId,
                systemId,
                records.size(),
                merkleRoot
        );

        // Reset for the next epoch
        epochId  = UUID.randomUUID().toString();
        records.clear();
        sequence = 0;

        return summary;
    }

    // -------------------------------------------------------------------------
    // EpochSummary
    // -------------------------------------------------------------------------

    /**
     * Immutable summary returned by {@link InferenceAuditor#closeEpoch}.
     *
     * <p>The {@code merkleRoot} is the root of a BRC-121 Merkle tree built from
     * the SHA-256 hashes of all records in the epoch.
     */
    public static final class EpochSummary {

        /** The UUID string that identified this epoch. */
        private final String epochId;

        /** The ARIA system identifier. */
        private final String systemId;

        /** Number of inference records in this epoch. */
        private final int recordCount;

        /**
         * Lowercase hex Merkle root (64 characters) computed over all record
         * hashes in this epoch.
         */
        private final String merkleRoot;

        public EpochSummary(String epochId, String systemId, int recordCount, String merkleRoot) {
            this.epochId     = epochId;
            this.systemId    = systemId;
            this.recordCount = recordCount;
            this.merkleRoot  = merkleRoot;
        }

        public String getEpochId()     { return epochId; }
        public String getSystemId()    { return systemId; }
        public int    getRecordCount() { return recordCount; }
        public String getMerkleRoot()  { return merkleRoot; }

        @Override
        public String toString() {
            return "EpochSummary{epochId='" + epochId
                    + "', systemId='" + systemId
                    + "', recordCount=" + recordCount
                    + ", merkleRoot='" + merkleRoot + "'}";
        }
    }
}
